### PR TITLE
Added gradient line drawing to LineChartRenderer

### PR DESCRIPTION
**Pull request by jjatie**
Reimplemented #649 for modern Swift and Charts

### CODE REVIEW

**Comment by natacodes** [comment]
Excited to see this re-implemented for a recent version of Charts. Thank you, @jjatie 👍🏻
Could you provide an example usage? I pulled your branch, added the following code but I don't get any gradient:
```
        set1.isDrawLineWithGradientEnabled = true
        set1.gradientPositions = [40.0, 120.0]
```

Also it might be a good idea to expose gradient colours as an open variable.

File: Source/Charts/Renderers/LineChartRenderer.swift [hunk 1]
```diff
@@ -751,4 +767,113 @@ open class LineChartRenderer: LineRadarRenderer
...
+        let generatedPath = CGMutablePath()
+        e = dataSet.entryForIndex(from)
+        if e != nil
+        {
+            generatedPath.move(to: CGPoint(x: CGFloat(e.x), y: CGFloat(e.y) * phaseY), transform: matrix)
+        }
+
+        // create a new path
+        let to = Int(ceil(CGFloat(to - from) * phaseX + CGFloat(from)))
+        for i in (from + 1)..<to
```
Comment (natacodes) [inline Source/Charts/Renderers/LineChartRenderer.swift]: `_xBounds.max` and `_xBounds.min` can be zeros this for is crashing the app in this case with `Fatal error: Can't form Range with upperBound < lowerBoun`

Comment (natacodes) [reply Source/Charts/Renderers/LineChartRenderer.swift]: It needs `if dataSet.entryCount > 1` somewhere.

**Comment by liuxuan30** [comment]
@natacodes thanks! good to have you help reviewing and providing feedback.
We plan this for 4.0 feature.

**Comment by jjatie** [comment]
@natacodes Thanks for the feedback. To be honest, I blindly reimplemented the solution from #649. I just wanted to get the ball rolling again on what I thought was a good feature. I would like to have an API that is unambiguous as to what values to provide as you mentioned.

I will hopefully have time in the next week or so to more closely evaluate and test the solution. In the meantime, feel free to PR any suggestions you have for the implementation into this branch.

**Comment by natacodes** [comment]
> I encourage you to PR any suggestions you have for the implementation into this branch.

@jjatie I tried, but it's a little bit too complex for me at this time.

**Comment by jjatie** [comment]
@natacodes No problem. This will take a little time for us to get around to, but I hope you are able to continue provide feedback when we do. Thank you.

**Comment by liuxuan30** [comment]
I actually wanted to try this out in my new app after we take care of others.

File: Source/Charts/Renderers/LineChartRenderer.swift [hunk 2]
```diff
@@ -751,4 +767,113 @@ open class LineChartRenderer: LineRadarRenderer
...
+            } else if (normPositionLocation > 1) {
+                gradientLocations.append(1)
+            } else {
+                gradientLocations.append(normPositionLocation)
+            }
+        }
+
+        for _ in dataSet.colors
+        {
+            cColor = dataSet.color(atIndex: 0)
```
Comment (larryonoff) [inline Source/Charts/Renderers/LineChartRenderer.swift]: @jjatie just want to check that this line is correct. It iterates over `dataSet.colors`, but always takes color at index `0`.

**Comment by larryonoff** [comment]
I think this PR can be closed now, since PR #3415 is based on this one and it was just merged into branch 4.0.0.

File: Source/Charts/Renderers/LineChartRenderer.swift (hunk 2 above)
Comment (jjatie) [reply Source/Charts/Renderers/LineChartRenderer.swift]: for i in dataSet.colors.indices
        {
            cColor = dataSet.color(atIndex: i)

**Comment by Ferszt** [comment]
Hey guys, can this functionality be used in production now?

**Comment by larryonoff** [comment]
@Ferszt please use branch #4.0.0. The functionality already there.

**Pull request closed without merging**